src/main/java/com/odysseusinc/arachne/portal/service/analysis/impl/AnalysisFilesSavingServiceImpl.java

```diff
@@ -0,0 +1,326 @@
+                    savedFiles.add(saveFile(f.getFile(), user, analysis, f.getLabel(), f.getExecutable(), null));
+                } else {
+                    errorFileMessages.add("Invalid file: \"" + f.getLabel() + "\"");
+                }
+            } catch (AlreadyExistException e) {
+                errorFileMessages.add(e.getMessage());
+            }
+        }
+        if (!errorFileMessages.isEmpty()) {
+            throw new ValidationRuntimeException("Failed to save files", ImmutableMap.of("file", errorFileMessages));
+        }
+        return savedFiles;
+    }
+
+    @Override
+    @PreAuthorize("hasPermission(#analysis, "
+            + "T(com.odysseusinc.arachne.portal.security.ArachnePermission).UPLOAD_ANALYSIS_FILES)")
+    public List<AnalysisFile> saveFiles(List<MultipartFile> multipartFiles, IUser user, A analysis, CommonAnalysisType analysisType,
+                                        DataReference dataReference) throws IOException {
+
+        List<MultipartFile> filteredFiles = multipartFiles.stream()
+                .filter(file -> !(CommonAnalysisType.COHORT.equals(analysisType) && file.getName().endsWith(OHDSI_JSON_EXT)))
+                .filter(file -> !file.getName().startsWith(ANALYSIS_INFO_FILE_DESCRIPTION))
+                .collect(Collectors.toList());
+
+        List<AnalysisFile> savedFiles = new ArrayList<>();
+        List<String> errorFileMessages = new ArrayList<>();
+        for (MultipartFile f : filteredFiles) {
+            try {
+                final boolean isExecutable = analysisService.detectExecutable(analysisType, f);
+                savedFiles.add(saveFile(f, user, analysis, f.getName(), isExecutable, dataReference));
+            } catch (AlreadyExistException e) {
+                errorFileMessages.add(e.getMessage());
+            }
+        }
+        if (!errorFileMessages.isEmpty()) {
+            throw new ValidationRuntimeException("Failed to save files", ImmutableMap.of(dataReference.getGuid(), errorFileMessages));
+        }
+        return savedFiles;
+    }
+
+    @Override
+    @PreAuthorize("hasPermission(#analysis, "
+            + "T(com.odysseusinc.arachne.portal.security.ArachnePermission).UPLOAD_ANALYSIS_FILES)")
+    public AnalysisFile saveFile(MultipartFile multipartFile, IUser user, A analysis, String label,
+                                 Boolean isExecutable, DataReference dataReference) throws AlreadyExistException {
+
+        ensureLabelIsUnique(analysis.getId(), label);
+        String originalFilename = multipartFile.getOriginalFilename();
+        String fileNameLowerCase = UUID.randomUUID().toString();
+        try {
+            Path analysisPath = analysisHelper.getAnalysisPath(analysis);
+            Path targetPath = Paths.get(analysisPath.toString(), fileNameLowerCase);
+
+            Files.copy(multipartFile.getInputStream(), targetPath, REPLACE_EXISTING);
+
+            AnalysisFile analysisFile = new AnalysisFile();
```

Comment:
`saveFile` and  `saveFileByLink` are both create new AnalysisFile(), is it possible to extract this code and make it common?